src/frontend/src/utils/templateRef.ts

```diff
@@ -0,0 +1,40 @@
+import { TemplateResult, render } from "lit";
+import { Ref } from "lit/directives/ref.js";
+
+/** This module contains some helpers that bring element lookup closer to the template description.
+ * By using "lit" `Ref`s, DOM elements can be bound without needing `id`s. Then, when the `TemplateRef`
+ * is rendered using `renderTemplateRef`, references are resolved returned.
+ *
+ * This means that, assuming references are given the proper types, element creation
+ * is completely type-safe.
+ */
+
+/** A helper type that represents a "lit" template that contains (typed) references. */
+export type TemplateRef<T> = {
+  template: TemplateResult;
+  keys: { [Property in keyof T]: Ref };
+};
+
+/** A wrapper around "lit"'s `render` that (blindly) reifies the template references */
+export function renderTemplateRef<T>(
+  tpl: TemplateRef<T>,
+  container: HTMLElement
+): T {
+  render(tpl.template, container);
+
+  // XXX: there may be a way to describe what's happening to tsc but I
+  // haven't found it.
+  // eslint-disable-next-line
+  const ret = {} as any;
+
+  for (const i in tpl.keys) {
+    const val = tpl.keys[i].value;
+    if (val === null) {
+      throw Error("Bad reference found in template");
```

Comment:
@peterpeterparker what do you reckon is the best way to handle errors?

Reply:
we have a beautiful assertion checker in `utils-js` if it's what you are looking for?

https://github.com/dfinity/ic-js/blob/main/packages/utils/src/utils/asserts.utils.ts

Reply:
No, I am looking for the best way to report the error. Here, if we do throw an error, then we lose track of all references. An alternative would be to not fail completely here and instead return e.g. useless elements.

Reply:
Don't really have a best approach to speak frankly, always depends a bit on how critical the error is I would say. Generally, maybe, when it's a DOM / components related error I try to handle it as close as I can. Like displayed an invalid entry in an input next to the input but, if it's a functional issue then it's often throw error and catched to be displayed in some sort of toasts or something I guess. Does that answer it?

Being said, try the assert of our utils, it's nice type safety.